controllers/module_nmc_reconciler.go

```diff
@@ -0,0 +1,150 @@
+//+kubebuilder:rbac:groups=kmm.sigs.x-k8s.io,resources=nodemodulesconfigs,verbs=get;list;watch;update;patch;create
+
+const (
+	ModuleNMCReconcilerName = "ModuleNMCReconciler"
+)
+
+type ModuleNMCReconciler struct {
+	kernelAPI   module.KernelMapper
+	filter      *filter.Filter
+	reconHelper moduleNMCReconcilerHelperAPI
+	nmcHandler  nmc.ModuleNMCHandler
+}
+
+func NewModuleNMCReconciler(client client.Client,
+	kernelAPI module.KernelMapper,
+	nmcHandler nmc.ModuleNMCHandler,
+	filter *filter.Filter) *ModuleNMCReconciler {
+	reconHelper := newModuleNMCReconcilerHelper(client)
+	return &ModuleNMCReconciler{
+		kernelAPI:   kernelAPI,
+		filter:      filter,
+		reconHelper: reconHelper,
+		nmcHandler:  nmcHandler,
+	}
+}
+
+func (mnr *ModuleNMCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// get the reconciled module
+	logger := log.FromContext(ctx)
+
+	logger.Info("Starting Module-NMS reconcilation", "module name and namespace", req.NamespacedName)
+
+	mod, err := mnr.reconHelper.getRequestedModule(ctx, req.NamespacedName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info("Module deleted")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get the requested %s KMMO CR: %w", req.NamespacedName, err)
+	}
+
+	// get all nodes
+	nodes, err := mnr.reconHelper.getNodesList(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get nodes list: %v", err)
+	}
+
+	errs := make([]error, 0, len(nodes))
+	for _, node := range nodes {
+		kernelVersion := strings.TrimSuffix(node.Status.NodeInfo.KernelVersion, "+")
+		mld, err := mnr.kernelAPI.GetModuleLoaderDataForKernel(mod, kernelVersion)
+		if err != nil && !errors.Is(err, module.ErrNoMatchingKernelMapping) {
+			logger.Info(utils.WarnString(fmt.Sprintf("internal errors while fetching kernel mapping for version %s: %v", kernelVersion, err)))
+			errs = append(errs, err)
+			continue
+		}
+
+		err = mnr.nmcHandler.HandleModuleNMC(ctx, mod, mld, node)
+		errs = append(errs, err)
+	}
+
+	err = errors.Join(errs...)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to reconcile module %s/%s with nodes: %v", mod.Namespace, mod.Name, err)
+	}
+	return ctrl.Result{}, nil
+}
+
+//go:generate mockgen -source=module_nmc_reconciler.go -package=controllers -destination=mock_module_nmc_reconciler.go moduleNMCReconcilerHelperAPI
+
+type moduleNMCReconcilerHelperAPI interface {
+	getRequestedModule(ctx context.Context, namespacedName types.NamespacedName) (*kmmv1beta1.Module, error)
+	getNodesList(ctx context.Context) ([]v1.Node, error)
+}
+
+type moduleNMCReconcilerHelper struct {
+	client client.Client
+}
+
+func newModuleNMCReconcilerHelper(client client.Client) moduleNMCReconcilerHelperAPI {
+	return &moduleNMCReconcilerHelper{
+		client: client,
+	}
+}
+
+func (mnrh *moduleNMCReconcilerHelper) getRequestedModule(ctx context.Context, namespacedName types.NamespacedName) (*kmmv1beta1.Module, error) {
```

Comment:
Isn't this function a dup of `getRequestedModule` in the `module_reconciler.go` file?

Reply:
As discussed in Slack, i am leaving it here and we can decide on it and the rest of functions in packages in v2.1